### ISSUE TITLE
Create official Java driver

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#3930]
https://github.com/npiv/rethink-java-driver might be a good jumping off point. It seems both of the 3rd party drivers for Java are abandoned or very out of date

**Comment by danielmewes** [comment]
I think we should try to get a beta release out by 2.1, especially since we might abandon protocol buffer support with that release and @npiv's driver is still using it.

Later we'll need to integrate it into our polyglot testing framework.

Another possible starting point for a driver is @dkhenry's Java driver. I've already ported it to the JSON protocol for benchmarking purposes in https://github.com/danielmewes/rethinkjava , but in contrast to @npic's driver it's missing support for lambdas.

**Comment by deontologician** [comment]
Most likely the serialization stuff is easier to port from one to the other. There are some language design issues to consider as well though, I think the default should return a cursor, rather than needing runForCursor. So in any case, there would be a lot of fiddling to get iit into shape

**Comment by kclay** [comment]
I've been meaning to add more javaness to my Scala driver(   https://github.com/kclay/rethink-scala) since it enforces type safety and provides methods for casting, blocking and async modes etc. On @deontologician note about the need for `runForCursor`. The driver is built in a way that its able to determine if the result should be a single value or an cursor. Pretty sure I can refactor some of this to work with java since some parts relying strongly on Scala implicits. Just a thought.

**Comment by stephanos** [comment]
@deontologician 
Will the new driver support Java 7?

**Comment by stephanos** [comment]
@danielmewes 
I can't even get your fork to compile! After some fixing I am stuck with `Error:(296, 37) java: incompatible types: long cannot be converted to com.google.protobuf.CodedInputStream` (RqlConnection.java) @ https://github.com/stephanos/rethinkjava

**Comment by danielmewes** [comment]
@stephanos I saw you did some patches in your branch, did you get it to compile eventually? I'm not sure why it's not compiling (though I'm not surprised, I've only ever used it on one type of system for a single purpose). My guess would be an incompatible protocol buffer library or compiler.

The official driver will not be based on that code base by the way.

**Comment by stephanos** [comment]
@danielmewes no, I didn't :( -will have to wait for the official driver.

**Comment by danielmewes** [comment]
@stephanos another driver you can try until then is https://github.com/dropbox/rethink-java-driver

**Comment by deontologician** [comment]
@stephanos I'm building it with Java 8, and we will probably release the beta version tested only with Java 8, but supporting Java 7 shouldn't be too hard and I think we should support it as part of release version

**Comment by deontologician** [comment]
Actually @stephanos what is your opinion on [the Java 7 EOL announcement](http://www.infoq.com/news/2015/05/Oracle-Ends-Java-7Public-Updates).

I assume there are lots of people still on Java 7, but I can't imagine everyone is able to switch right away. So as someone using Java 7, I'm curious whether upgrading is on your roadmap or what the feeling is towards Java 8 from professional Java developers in general

**Comment by stephanos** [comment]
@deontologician Well, I'd (almost) give a kidney to be able to use Java 8 right now. Problem is that we run on App Engine which only supports Java 7 - and it looks like this won't change, ever. The new option is ManagedVM, but it's only in beta status, has no SLA and is super slow. So we're stuck on Java 7 :(

**Comment by deontologician** [comment]
@stephanos thanks, that's really useful info

**Comment by iboumedien** [comment]
Thanks

**Comment by scottwleonard** [comment]
@deontologician I wanted to add my $.02 to your question about Java 8. We have been using Java 8 for awhile now, and there are a ton of useful features. streams are very well implemented from a syntax perspective, and they have allowed us to write our code in a more functional way. We didn't have any issues converting our codebase over from Java 7.

**Comment by deontologician** [comment]
Thanks @scottwleonard, that's helpful. I'm a fan of Java 8, I think (despite being Oracle) that they did a good job integrating functional language features into the existing language in a way that is really nice to work with.

**Comment by gmethvin** [comment]
I'd also like to see good Scala driver support but once the low level stuff is officially supported in a Java driver I think it'd be fairly easy to write a small Scala layer on top of it with a nice DSL and maybe some additional type safety.

For the Java driver, ideally the serialization framework should be pluggable, and it'd be good if cursors implemented reactive streams so they'd easily interoperate with Play iteratees and Akka streams.

**Comment by danielmewes** [comment]
@gmethvin Thanks for the suggestions.

I think we'll first make sure that we ship a Java driver on par with our other official drivers and with a comparable interface that will be based on synchronous cursors.
Once we have that working reliably, I suspect that adding a reactive stream API is going to be fairly easy from what I can tell (I had only a very brief look at https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.0/README.md , so there might be some complications that I missed).

@deontologician do you have additional comments on that?

@gmethvin What is your use case for a pluggable serialization framework? Are you talking about the JSON encoder/decoder, the connection logic, or some higher-level API?

**Comment by deontologician** [comment]
I don't have much to add except that it will take me a bit of research to find out what the best approach is for async etc. The Java ecosystem is really big, so it might make sense to do something like the python driver where the async library is swappable.

**Comment by gmethvin** [comment]
@danielmewes I'm talking about the JSON serialization. It would be nice if I'm already using Jackson, Gson, Play Json, etc., to easily write an adapter to serialize/deserialize my objects.

**Comment by danielcompton** [comment]
I maintain the Clojure RethinkDB library, and would be very interested in using an official RethinkDB Java driver for connection management and cursors, and wrapping/adapting this to be idiomatic in Clojure.

**Comment by gmethvin** [comment]
Also, the type system should try to prevent you from doing things that are obviously wrong. It shouldn't be possible to have a runtime error like ` Indexed order_by can only be performed on a TABLE or TABLE_SLICE` since the type system prevents you from calling the method.

**Comment by danielmewes** [comment]
I'm quite sceptical about putting anything but trivial type constraints into the driver.
It adds additional code and increases the maintenance burden, since we would need to update the driver when adding new type signatures to ReQL.

**Comment by coffeemug** [comment]
Another reason to avoid strict type checking in the driver is that the ReQL type system is quite complex, and it would be very hard to encode it in Java's type system. For example, currently `orderBy` takes an optarg for the index; the output type would depend on whether the optarg is present. Doing that sort of stuff well is really hard; we could break up `orderBy` into two functions, but that's just one of many, many examples where there is a gap between ReQL's type system and Java's.

**Comment by deontologician** [comment]
OK, I am working on this again now that the webui stuff for 2.1 is mostly done

**Comment by hepin1989** [comment]
mixing the Reactive-Stream and the rethinkdb's realtime feature  would be very cool.and the java driver I think should be total async,I suggest to build it on top of netty4.
once the java driver is release,I will migrate our game server cluster to use rethinkdb,waiting too long.

**Comment by gmethvin** [comment]
I agree with @hepin1989 that the driver should be async by default. Making it implement a reactive stream interface would be nice and is probably worth it since it'll make it easy for rethinkdb to interoperate with other reactive libraries.

**Comment by deontologician** [comment]
OK, the version of the driver on branch `josh/java-driver` pretty much works. You'll need to `gradle build` in the `drivers/java` directory. There's no documentation or nice packages yet, so only try it out if you're feeling plucky. We'll have the packaged beta version soon.

**Comment by deontologician** [comment]
Note: consider this the unofficial "alpha" for people following this thread. There should be a decent number of bugs still, but the major work is done

**Comment by scottwleonard** [comment]
thanks @deontologician very excited about this

**Comment by jaanek** [comment]
Great job @deontologician . I'm currently looking mainly to insert & watch table changes using the java driver (real-time capabilities) but did not notice how I can asynchronously watch table changes using the current driver in josh/java-driver branch.

**Comment by deontologician** [comment]
Yeah, async is on the list. Once I run it though out full language test suite it will be in better shape to add features like that

**Comment by dklotz** [comment]
Really looking forward to this! Just out of curiosity: Will this driver implement some form of connection pooling itself, or would it need to be done on top of the driver?

**Comment by deontologician** [comment]
There is no connection pool for now, but we have plans to add a pool to all of the drivers eventually

**Comment by hepin1989** [comment]
netty just add the connection pool,maybe we could build on top of it.

**Comment by hepin1989** [comment]
@deontologician seems you build directly on NIO and required java8.

**Comment by deontologician** [comment]
@hepin1989 would you mind opening a new issue with your suggestion? I did the io in the most straightforward way for the first version, but if there's a better way that will integrate better I'm open to it.

**Comment by jaanek** [comment]
@deontologician Could not figure out currently on how to insert a record to a table. If using it like this: r.table("test").insert("{ name: 'Star Trek TNG 2', episodes: 177 }").run(conn) then it gives an error: "Cannot convert STRING to SEQUENCE". Could you give a quick example on how to do it using current driver?

**Comment by deontologician** [comment]
@jaanek, you have to insert a map, so it would need to be ```java
r.table("test").insert(new MapObject()
    .with("name", "Star Trek TNG 2")
    .with("episodes", 177)
).run(conn);
```

**Comment by deontologician** [comment]
The MapObject is a convenience from the original npiv driver, but any `Map<String,Object>` will work

**Comment by jaanek** [comment]
Thanks @deontologician . But this gives me this "StackOverflowError":
```
Exception in thread "main" java.lang.StackOverflowError
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:64)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:29)
	at com.rethinkdb.model.Arguments.coerceAndAdd(Arguments.java:43)
	at com.rethinkdb.model.Arguments.<init>(Arguments.java:18)
	at com.rethinkdb.gen.ast.MakeObj.<init>(MakeObj.java:20)
	at com.rethinkdb.gen.ast.MakeObj.fromMap(MakeObj.java:33)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:71)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:29)
	at com.rethinkdb.model.Arguments.coerceAndAdd(Arguments.java:43)
	at com.rethinkdb.model.Arguments.<init>(Arguments.java:18)
	at com.rethinkdb.gen.ast.MakeObj.<init>(MakeObj.java:20)
	at com.rethinkdb.gen.ast.MakeObj.fromMap(MakeObj.java:33)
        ...
        at com.rethinkdb.ast.Util.toReqlAst(Util.java:71)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:29)
	at com.rethinkdb.model.Arguments.coerceAndAdd(Arguments.java:43)
	at com.rethinkdb.model.Arguments.<init>(Arguments.java:18)
```

**Comment by deontologician** [comment]
Ah, OK that would be a bug then :) I'll check it out tuesday

**Comment by hepin1989** [comment]
@deontologician https://github.com/rethinkdb/rethinkdb/issues/4802 I have opened one as you asked:)

**Comment by deontologician** [comment]
@jaanek I've fixed the recursion bug on the branch

**Comment by jaanek** [comment]
Thanks @deontologician, it works.

**Comment by eric-poitras** [comment]
Hi,
it is useful for Connection to be a Generic type ? Java can be a bit strange in the way he handle generics and sometime it's just better to avoid it. Here a sample case of what I am talking about:

```java
public class SomeClass implements SomeApi {

    private final Db         db;
    private final Connection conn;

    @Inject
    public DeviceManagerImpl( Db db, Connection conn ) {
        this.db = db;
        this.conn = conn;
    }

    @Override
    public List<Device> listDevices() throws ApiException {
        Cursor<Map<String, Object>> c = db.table( "test" ).run( conn );
    }
```

The line: Cursor<Map<String, Object>> c = db.table( "test" ).run( conn ); won't compile because I did not defined the variable conn as a generic type Connection<?> then java call the "run" method in JDK 1.4 style so it returns Object instead of <T>. (By the way, IntelliJ don't even highlight this as a problem).

Anyway, it's not a big deal, just an annoyance. You fixed the stack overflow and thinks are looking better now. Keep up the good work !

**Comment by deontologician** [comment]
I added the generic parameter to match the way the python driver handles connections, but in hindsight I don't think it'll end up being useful. I've run into the issue you've mentioned a few times and I think I agree

**Comment by sepbot** [comment]
Can't seem to get any results with the getAll method using secondary keys. The following two work:
```java
getAll("value")
getAll("value", "primary key")
```
But the following doesn't return any results (equivalent js in Data Explorer returns results):
```java
getAll("value", "secondary key")
```
Are you already aware of this? Could I be doing something wrong? Should I file a bug report?

**Comment by deontologician** [comment]
Ah, you need to use the optarg, so `getAll("value").optarg("index", "secondary_index") `

**Comment by danielcompton** [comment]
Would it be possible to start publishing SNAPSHOTs of this to Maven Central so people can start experimenting with it and using it as a dependency?

**Comment by deontologician** [comment]
Right now I'm focused on getting our tests to run to ensure the driver is complete. I'm not familiar with what's required to post it up, its on my to-do list. I would really love to do it though, if you guys who do this all the time could submit a PR or maybe just detail what I need to do to get it published so I don't have to take time out it'd be awesome.   

Either way though, it's coming, I just have to do this stuff one step at a time :)

**Comment by dandc87** [comment]
Sonatype (the Maven Central proprietor) has great documentation on how to create an account and configure a project to upload builds

http://central.sonatype.org/pages/producers.html
http://central.sonatype.org/pages/ossrh-guide.html
http://central.sonatype.org/pages/gradle.html

**Comment by mattias800** [comment]
I am experimenting with the Java driver, and I though I should mention that it has a dependency.

Just add to pom.xml:

        <dependency>
            <groupId>com.googlecode.json-simple</groupId>
            <artifactId>json-simple</artifactId>
            <version>1.1.1</version>
        </dependency>

**Comment by deontologician** [comment]
The Java driver doesn't use maven directly, it uses gradle, so there is no pom.xml

**Comment by mattias800** [comment]
Yes I know, but if you build it with Gradle, and install the resulting JAR in your local Maven repo and try to use it in a project, you will get an exception. Add json-simple as a dependency and it will work.

Also, conn.use(dbName) doesn't work currently, you have to use r.db(dbName).table(..).

Not sure if you're looking for bug reports or not, but I thought I'd mention it :)

I'm really looking forward to an official release of the driver, we're looking to migrate from MySQL to RethinkDB, using Java. Is there an ETA on a release (beta or not) yet?

**Comment by laguiz** [comment]
Very excited by this driver. It makes RethinkDB a lot more appealing for us (Mainly Java backends).

**Comment by deontologician** [comment]
It's coming soon :tm: . I'm opening a sonatype account and getting packaging ready etc. All tests (that aren't being skipped) are passing with the code in `next`, so it's on its way definitely. I'll check out the `conn.use()` bug as well.

**Comment by pires** [comment]
Looking at all the options, `dropbox` fork seems to be the maturest. But, since it's not available on Maven Central nor is it supported, I am eager to contribute to the official driver.

**Comment by deontologician** [comment]
OK, I have finagled uploading a snapshot release to https://oss.sonatype.org/content/repositories/snapshots/com/rethinkdb/rethinkdb-driver/ if anyone would like to try it out and tell me if they have any problems with it. Gradle auto-generates the pom.xml required, so it should solve the issue @mattias800 was having with undeclared dependencies

**Comment by deontologician** [comment]
@mattias800 I fixed the `conn.use` problem on the current branch

**Comment by mattias800** [comment]
Nice! I'll try it out!

Maybe Java support is in an experimental, unofficial phase so far, but are there any plans on supporting RxJava? That would be really nice, and would fit nicely with the changes() feed.

Making the SDK non-blocking is nice regardless :)

**Comment by deontologician** [comment]
RxJava or Reactive streams would be neat, but we don't have an issue open for them currently. Feel free to open one, and tell us what you'd like to see

**Comment by hhimanshu** [comment]
Do we have this driver officially out?

**Comment by deontologician** [comment]
@hhimanshu not yet, but very soon. There is a snapshot available in maven central though

**Comment by crockpotveggies** [comment]
Guys just wanted to confirm this is the Java Driver that's sitting up in the Sonatype snapshots repo? https://github.com/rethinkdb/rethinkdb/tree/josh/java-driver/drivers/java

**Comment by crockpotveggies** [comment]
Also, would someone be nice enough to post the javadoc since http://rethinkdb.com/api/java is 404?

**Comment by igorlukanin** [comment]
@crockpotveggies Yup, what you see in the Sonatype repo is the current snapshot of Java driver.

As far as I know, there're no Javadocs now. You can have a look [at the tests](https://github.com/rethinkdb/rethinkdb/blob/josh/java-driver/drivers/java/src/test/java/RethinkDBTest.java) though. I hope it helps.

**Comment by crockpotveggies** [comment]
Awesome thanks! Last question if you don't mind answering (I don't know where else to ask). I've already got a test integration working with a project of mine. How does the `.filter` function work? I tried doing something like `.filter(new Javascript("""{"userId":"Mary"}"""))` with no dice. I couldn't find any examples in the test spec.

(Note: the triple-quote is a Scala-ism)

**Comment by igorlukanin** [comment]
@crockpotveggies I think that the best way to filter is to use a lambda. Please have a look [at the test in this PR](https://github.com/rethinkdb/rethinkdb/pull/5053/files).

As for the ways to communicate on drivers issues, it may be useful to check the official RethinkDB Slack channel. You may [sign up in a few seconds](http://slack.rethinkdb.com). There's even the `driver-dev` channel there :)

**Comment by crockpotveggies** [comment]
@igorlukanin awesome thanks for your help!

**Comment by coffeemug** [comment]
FYI, the driver hasn't yet been officially released (though it will any day now). There will be comprehensive API docs when it's out; writing those is non-trivial and @chipotle deserves an Oscar for the work he's currently doing. Sorry the docs are currently missing; this is very temporary.

**Comment by crockpotveggies** [comment]
@coffeemug not a problem reason I'm poking around now is I'm launching a new project in two months and I wanted to stick with RethinkDB as in my original planning. Totally fine if I have to poke around the source code and work with limited docs :)

**Comment by chipotle** [comment]
For what it's worth, you can monitor the java driver documentation on its branch:

https://github.com/rethinkdb/docs/tree/901-java-driver

The API documentation is largely done in draft but it's not fully reviewed, so some pieces of it may turn out to be horrible lies. (A lot of the stuff about changefeeds may be wrong and the cursor section definitely is, although we'll be straightening it out.)

You can also look at the comments (and discussion) on the docs issue:

https://github.com/rethinkdb/docs/issues/901

That has some useful quick overviews of Java driver quirks, notably the differences in the way the `connect` command works and the way optional arguments are passed with the `optArg` command. The API docs also talk about some Java-specific helper functions like `r.array(1, 2, 3, ...)` to create arrays and `r.hashMap("field1", "value1").with("field2", "value2")` to create hashmaps.

**Comment by tyth** [comment]
I toyed with Snapshot version yesterday (in Scala codebase first, then checked if it is any different in Java code) and my experience was rather negative. The API design principles just don't fit well with static typed languages.

There no information in code about what types of parameters are required, there is (worst part) no information about what type will be returned. I have to find out, from documentation or library source code, what the type is and explicitly declare it. And it can go wrong any time.
The great benefit of static typing - cheching things on compile time is lost. 

Seriosly, when type checked SQL in Java/Scala community widely adopted and we have all these awesome DSLs for non relational databases, this kind of approach is terrible.

**Comment by coffeemug** [comment]
@tyth -- do you think it's a general consequence of building a DSL in a type-checked language, or do you think it's a result of idiosyncrasies of this particular driver/design?  What do you think we could do to make the driver better?

**Comment by mattias800** [comment]
I'm curious as well.

I could see it automatically populating (and returning) data objects given a class as a parameter (like GSON does with deserialization). But I'm not sure how you would add typing to the communications API?

Having it return instances of your own classes would be a great feature though.
See rxjava-jdbc for an example: https://github.com/davidmoten/rxjava-jdbc

```
Observable<Person> persons = db
                 .select("select name, score from person order by name")
                 .autoMap(Person.class);
```

And you could use (preferably standard) annotations for mapping properties to class members.

```
public interface Person {

    @Column("name")
    String name();

    @Column("score")
    int score();
}
```

**Comment by tyth** [comment]
@coffeemug I see that most of the code is generated metajava.py. I haven't read it, so there could be no possible way of meeting some requirements of "nice" static-typed lib.

But what I generally want:
1. Type boundaries on method parameters (ex:  how can `r.table(Object tableName)` method take any `Object`? Why can't we state that it can be only a `String`?)
2. Sane return types. So `tableList.contains(table)` would return `Boolean`, `table.changes` would return `Cursor<Map ...>` etc.

Mapping to classes, @mattias800 writes about, is a great thing too, but it can be always added later on, either in a standard driver or as a separate library. 

I would be glad to contribute (e.g. async, intergration with streams, Scala wrapper later on), but so far as code is generated in such peculiar (for Java) way, I don't see how I can do it.

**Comment by mattias800** [comment]
I would much rather see RxJava integration (using Observable), than streams.
Other than that, I agree completely :)

**Comment by kclay** [comment]
I've been working on the scala driver for past 2 years or so.. Its strictly typed so it may help in providing a guide on some of the quirks I ran into. Just a note group/ungroup may not be possible to type with java. I had to go thru some tricky scala setup to half way get it to work.

https://github.com/kclay/rethink-scala

I've been meaning to relax some of the code to provide a java version but since the official driver is being made maybe someone can try porting over some of the typed stuff. But for the most part most of the typed code can be ported as is.

I started to implement some streams a bit (scalaz-streams) but will be looking into adding other support (RXJava/Scala and alike).

Funny you mention the dsl @tyth, I actually had an issues for reworking with the official driver, kclay/rethink-scala#46. Maybe somehow we solve both of theses.

**Comment by hepin1989** [comment]
@mattias800 I would like Reactive-Stream implemented ,with this way you could integrate it with all other RS implement ,includs RXJava,and Flow is in Java 9 also.

**Comment by mattias800** [comment]
@hepin1989 Ah, yes, that would be great!

**Comment by deontologician** [comment]
@tyth 
> Type boundaries on method parameters (ex: how can r.table(Object tableName) method take any Object? Why can't we state that it can be only a String?)
This is due to the flexibility of ReQL. It's not that `r.table` accepts a string, it's that it accepts a term that evaluates to a string. For example:

```java
r.table(r.expr("Foo-").add(r.uuid()))
```
is a valid expression. In general, since the type of a ReQL query may depend on the data in your database, and we have no guarantees about the type or structure of that data, only a subset of ReQL can be statically typed, and my guess is you would be disappointed by how small that subset is.

We actually have a similar issue with return types:

> tableList.contains(table) would return Boolean

It actually doesn't return a Boolean, it returns a reql expression that evaluates to a boolean. We'd need to add some tricky generic stuff to our expression types to pass around the "expected return type".  So `.contains(foo)` would return a `ReQLExpr<Boolean>`. You could conceivably get pretty far doing it this way. There are tricky bits around `r.js`, and `.coerceTo` that would need special handling, and I expect some parts of ReQL would just not work. This isn't a basic kind of driver though, it's a pretty big undertaking, and right now it's beyond the scope of what the Java driver is attempting to do.

There is definitely a place for an ODM or a 3rd party driver to either allow you to make additional assumptions about the schema (with the ODM) or to restrict the driver to ReQL expressions that can be statically typed. But the official driver aims to cover all of ReQL, and it also aims to be maintainable, so I didn't go with the fancy typing route or restricting ReQL. 

(As an aside, to my knowledge, the only legal expression you can create with other drivers that you can't with the java driver is something like `r.branch(foo, bar, r.db("dbname"))` where a branch returns a database instead of a ReQLExpr)`

@mattias800 
> I could see it automatically populating (and returning) data objects given a class as a parameter (like GSON does with deserialization). But I'm not sure how you would add typing to the communications API?

Good news! Thanks to @igorlukanin's hard work in #4889 this is possible (I don't remember if this made it into the snapshot, but I'll upload it again later today to be sure). There's no docs yet, but you should be able to do:

```java
MyClass foo = someReqlExpr.run(conn, MyClass.class);
```

And it will do the coercion for you. For cursors, (I believe) it looks like:

```java
Cursor<MyClass> foo = someReqlExpr.run(conn, MyClass.class);
```

**Comment by igorlukanin** [comment]
@mattias800 There're also [tests with usage examples](https://github.com/rethinkdb/rethinkdb/blob/josh/java-driver/drivers/java/src/test/java/RethinkDBTest.java#L247) for the POJO stuff mentioned by @deontologician.

@deontologician I bet @tyth would really like if we make `r.table(...)`'s signature not just `r.table(Object tableName)`, but provide a more strongly typed `r.table(ReqlExpr tableNameExpr)` among with probably more frequently used `r.table(String tableName)` which would just wrap the string with the appropriate `ReqlExpr`.

I think there's no harm in doing this in means of maintainability, and it could be done anytime after 2.2 (kind of 1.0 for Java driver) is released.

**Comment by mattias800** [comment]
Nice! Gonna try it out tomorrow!

**Comment by pires** [comment]
@crockpotveggies and everyone else, just be aware that the artifact available on OSS Sonatype atm is dated Oct 26th and is outdated when compared to current code. Better build manually and install it locally.

**Comment by dklotz** [comment]
@mattias800 
> Having it return instances of your own classes would be a great feature though.

This (in essence: ODM / object-document-mapping) could be compared to e.g. what Morphia is doing for MongoDB on top of their normal java driver and I would strongly suggest to have it in a library separate from the basic driver. Definitely needed, but definitely a different beast than the basic DB driver.

**Comment by deontologician** [comment]
OK guys, I uploaded a new snapshot. We've also merged the driver to next (the main development branch) so things are coming along!

**Comment by crockpotveggies** [comment]
Nice work! Unless someone found a fix, I think compatibility with Scala
`case class` is currently broken, so not sure I can test the new version.
It will require some thought around reflection since case classes are
structured a little differently.

**Comment by pires** [comment]
@deontologician :clap:

**Comment by radicaled** [comment]
I've noticed that the POJO mapping code intentionally ignores fields that are inherited.

So, simple hierarchies like this fail:

```java
class Model {
  public String id;
}

// id will always be null if you pass this class, like so: `...run(connection, Person.class)`
class Person extends Model {
   public String name;
}
```

Is there any particular reason why the POJO mapper does this?

**Comment by pires** [comment]
I've also noticed it doesn't support `Enum`.

**Comment by kclay** [comment]
I would suggest  looking into Jackson for POJO support there are a lot if corner cases as well if the intended use was to also be able to support Scala case classes as well. I ended up using Jackson for the rethink-scala library and it worked out pretty well. 
On Nov 21, 2015 6:12 AM, Paulo Pires <notifications@github.com> wrote:I've also noticed it doesn't support `Enum`.

**Comment by deontologician** [comment]
@kclay you may be right. I opened #5128 to look at what would be required

**Comment by triggetry** [comment]
When I try to call insert in parallel I noticed that the driver crashes with a "java.lang.OutOfMemoryError"  Exception. This happens as soon as more than one calls to insert are fired in parallel. Any suggestions how to fix this?

Here is the stacktrace:

java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
Caused by: java.lang.OutOfMemoryError: Java heap space
	at com.rethinkdb.net.Util.leByteBuffer(Util.java:25)
	at com.rethinkdb.net.SocketWrapper.recvall(SocketWrapper.java:100)
	at com.rethinkdb.net.ConnectionInstance.readResponse(ConnectionInstance.java:74)
	at com.rethinkdb.net.ConnectionInstance.readResponse(ConnectionInstance.java:58)
	at com.rethinkdb.net.Connection.runQuery(Connection.java:156)
	at com.rethinkdb.net.Connection.run(Connection.java:204)
	at com.rethinkdb.ast.ReqlAst.run(ReqlAst.java:66)

kind regards,

**Comment by bchavez** [comment]
When I ported over the **Java** driver code to **C#**, I noticed some internal data structures were not not thread-safe in **C#** land. All the threading issues are resolved **C#** driver & documented in this internal design note:

https://github.com/bchavez/RethinkDb.Driver/issues/15

Although, this design note is for C#, it might help identify thread-sensitive spots in the Java driver since both drivers share the same architecture.

**Comment by danielmewes** [comment]
@triggetry I think @bchavez is right that this is likely a threading problem. For now I recommend using one connection per thread.

**Comment by danielmewes** [comment]
Opened https://github.com/rethinkdb/rethinkdb/issues/5166 to add thread-safety to the Java driver.

**Comment by redi-russela** [comment]
It's released, yay!

**Comment by a1russell** [comment]
It's released! Yay!

**Comment by deontologician** [comment]
This is fixed now.

**Comment by mglukhovsky** [comment]
Great work, @deontologician! :coffee:

**Comment by crockpotveggies** [comment]
Just wanted to say nice work to everyone who got this developed. Not sure if Scala support made it for the release, however it's very meaningful to the JVM community :)
